### PR TITLE
AppD consolidated update PR

### LINKED ISSUE #369: Add  categories metadata to AppD application definitions
## Enhancement Request
Add ~tags~ categories metadata to AppD application definitions.

### Use Case:
A large AppD will have applications in it for a variety of use-cases. To help users navigate and search the AppD metadata fields should be provided to classify applications, tags (keywords) are a convenient and flexible mechanism for doing so that can be used to implement a number of different search or navigation use-cases. Tags are not restrictive as to the meaning of the tags or number that can be applied to an application, ensuring that they can be used to sub-divide the directory in a number of different ways depending on the preferences of the owner of the AppD.

For example: 
> A user wishes to search for applications for working with Fixed Income orders, they can search the UI provided for their AppD implementation via keywords "Fixed Income orders" and review apps returned by search, or they can use the provided navigation to click on the "Fixed Income" tag and review associated apps, or they can use a provided filtering mechanism and select both the "fixed Income" and "orders" tags to filter the list of applications.

Proposed format:
```JSON
{
  "appId": "string",
  "name": "string",
  "manifest": "string",
  ...
  "categories": [
    "Fixed income"
    "Orders"
    "Anther tag"
  ],
  ...
```

### LINKED ISSUE #370: Improve the AppD's search function and documentation
## Enhancement Request 
It is not currently clear from the documentation of the AppD search function what query parameters that it is intended support. The search function could/should support filtering of the app catalog by:
- name
- title
- description
- tags
- publisher
- keyword (a combination of title, description, tags, name, publisher - this could replace all other search filters above, apart from perhaps name & tags)

It should not support:
- search via AppId (as this is redundant with the Get call to retrieve an application definition `/v1/apps/{appId}`

In the documentation (https://fdc3.finos.org/schemas/1.2/app-directory#tag/Application/paths/~1v1~1apps~1search/get) it appears that the Query parameters area in the docs is being used to provide definitions of the fields returned for each application. The intended query parameters should be clarified and the field descriptions re-written to indicate how each field should be handled in the search.

### Use Case:
Searching for applications is a common use-case that should require no explanation.  Standardizing the search parameters will ensure that different UI implementations can correctly call the search functions of AppD's they were not developed or tested against. This is not guaranteed based on the current specification and documentation

### Current (confusing) documentation of query parameters
image

### LINKED ISSUE #553: Add backwards compatibility support for the App Directory
PR #437 (issue #314) introduced a breaking change to the App Directory specification, namely changes to the `Application` schema. This issue suggests adding backwards compatibility to the specification by using the versioned URL end-points:
- Introduce `v2` end-points for all newly added changes
- Add back the old Application schema and sort out any component naming issues
- Mark the `v1` end-points as optional/deprecated

### LINKED ISSUE #555: Improve the description of AppD, its role and value.
### Area of Issue
[x] App Directory

### Issue Description: 
The description of the AppD in the FDC3 website is poor and fails to properly outline its role in FDC3, intended use and ultimately its value as part of the standard, potentially leading to recent proposals to remove it from the standard (#551 #550). 

The description should be improved to the point that an individual new to FDC3 can immediately understand its role and why they might want to use it (and/or encourage vendors they work with to use it).

The expected interactions between Desktop Agents and AppD should also be clarified as some readers have interpreted the presence intent configuration data in the appD as a requirement for intent resolution to involve queries to the appD during resolution or for the appD to control intent resolution, when it is actually only intended to provide data to support intent resolution, where implementing resolution is the province of the Desktop Agent implementation (#551).

### LINKED ISSUE #556: Better describe relationship between AppD definitions and other manifests
### Area of Issue
[x] App Directory

### Issue Description: 
The use-cases for the App Directory Standard differ from that of app definition schemas, such as the Web Application Manifest, in that AppD is intended to provide access to and a means to discover details of multiple applications (some or all of which will exist on different domains to the directory itself), and can include details of how to launch them in multiple different environments. Whereas, a web application manifest is hosted on the same domain as the application, targets web browsers (which sometimes have different needs/form factors from containers or other environments that might implement a Desktop Agent) and are normally only referenced from documents on the same origin (see https://www.w3.org/TR/appmanifest/#start_url-member) allowing the application to be 'installed', once it has been discovered. 

Since PR #437 (issue #314) was merged, the AppD's application definition can both specify the minimum level of detail need for launching a web app (via the `details.url` property) and can either embed or refer to (via URL) multiple other application manifests (via the `hostManifests` property) necessary for launching the application in proprietary containers and for launching native apps etc.. This relationship between the AppD's own definition format and directory and third party definition formats is not yet well described in the AppD's documentation, which should be improved to do so. 

### Additional Context:
- #314 
  - #437

### LINKED ISSUE #558: Correct "trusted identity" statements in AppD overview, consider whether app authentication properties are needed
### Area of Issue
[x] App Directory

### Issue Description: 
The App Directory Overview makes the claim that AppD: 

> "...provides trusted identity for financial desktop apps." 

However, this claim is hard to support under the current specification as the AppD doesn't obviously provide a lot of value in: 

> "preventing spoofing and man-in-the-middle attacks when apps communicate with one another and exchange data"

particularly where that exchange happens anonymously over context channels. The Application definition does not itself provide a mechanism to authenticate individual applications.

However, when an appD is accessed via HTTPS, SSL & DNS confirm for us that we are retrieving app definition data from the expected source. If we trust that source AND it is authoritative for the application (e.g. a domain owned by a vendor that publishes the app, or by one's own firm pointing to their own deployments of in-house or vendor apps), we can then use HTTPS URLs to launch and communicate with the (web) application, secure in the knowledge we are protected from spoofing or MITM attacks and are accessing the expected application.

Hence, there are ways that an AppD can provide a trusted identity (for a web app) - but the statement in the AppD overview is far too broad and doesn't acknowledge the above caveats and should be corrected.

Further, it should be considered whether the AppD's application definition schema should include properties that can be used to authenticate an application (for example a public key that could be used, in conjunction with an API call or other suitable mechanism) to verify an application's identity. Such a mechanism would also allow the AppD to provide a trusted identity for native applications that are not accessed via SSL/DNS.

### LINKED ISSUE #560: Define a term in the FDC3 glossary that can be used to refer to the application definition format used by the App Directory
### Area of Issue
[x] App Directory

### Issue Description: 
The FDC3 App Directory standard incorporates both a definition of a REST API for interacting with a directory of applications and format for defining each individual application. Define a term in the FDC3 glossary that may be used to refer to the application definition.

### LINKED ISSUE #561: Better align Application definition to Web Application Manifest and supplementary application information
## Enhancement Request
The FDC3 AppD application definition bears some similarity to the W3Cs draft [Web Application Manifest specification](https://www.w3.org/TR/appmanifest/) and supplementary properties intended to support digital storefronts in the[ Web App Manifest - Application Information specification](https://w3c.github.io/manifest-app-info/). 

Despite the fact that the use-case for the AppD differs from the Web Application Manifest and that it can include a reference to an external manifest (embedded or via a URL), there may be advantages to aligning the format of the AppD app definition with the Web Application Manifest as much as possible (e.g. an AppD app definition might be more easily created by importing an existing web app manifest). Further, inspiration can also be drawn from the web application manifest for ways to improve the appD app definition. 

For example:
- AppD's icons property is missing the purpose field
  - it may make sense to refine or extend the list of defined purposes
- Include Text direction and language properties
- Include a Categories property that may be used to organize a storefront and enhance search (see issue #369)
  - A list of categories relevant to financial services applications, rather than general web applications, could also be defined
- The AppD's images property could be renamed screenshots and extended to include the sizes, platform and description properties
  - New values for the platform property may need to be defined.

### Use Case:
- If field names and formats are aligned between the Web Application manifest and the App Directory an AppD app definition might be more easily created by importing an existing web app manifest). 
- Inspiration can also be drawn from the web application manifest for ways to improve the appD app definition.

### LINKED ISSUE #562: Add vendor-agnostic properties to the AppD application definition format for launching native apps
## Enhancement Request
Add vendor-agnostic properties to the AppD application definition format for launching native apps

FDC3 and the AppD are expected to support native and hybrid applications in addition to web apps. However, the appD app definition format currently requires a hostManifest (in an unstandardized, vendor-specific format) in order to support launching of native applications. 

Since the merge of PR #437, the appD's application definition format includes a vendor agnostic `details.url` property that can be used to launch web applications. Consider the addition of further sub-properties of `details` that will support vendor-agnostic launching of native applications. 

### Use Case:
Vendors of native apps, and in-house native app development teams may wish to publish details of their apps for inclusion in app directories. Despite the fact that native apps may not be launchable without installation, there are several use-cases for inclusion in an appD:
- Providing configuration to customers that have installed the applications and wish to use them with FDC3 (i.e. to save the customer figuring an optimised configuration themselves).
- Native apps that do support launch without installation (e.g ClickOnce apps).
- In-house developed native apps that are already deployed.
- Supporting discovery of apps that could be installed to work with FDC3

### LINKED ISSUE #565: Make it possible to retrieve all applications in an AppD
## Enhancement Request

A common use case for the AppD is providing a list of all available applications to a desktop agent. However, there currently exists no endpoint for retrieving a listing of all applications. AppD implementations may currently be providing this feature by supporting calls to the search endpoint with no parameters, however, this feels contrary to the purpose of that endpoint (to allow searching of filtering and the return of a subset of applications). 

Add support for an endpoint that returns all application details, e.g. `GET /<version>/apps`

### Use Case:
- Retrieving details of all applications available in an AppD (to the logged-in user if authentication is in use) in order to populate a launcher menu or app catalog user interface.

### LINKED ISSUE #566: Endpoints for adding an app to an appD will often be implementation dependent and should not be required for compliance
### Area of Issue
[x] App Directory

### Issue Description: 
The AppD spec currently defines a POST endpoint (`/<version>/apps`) for adding an application to an appD, and in the absence of any clarifying info, this is currently required for compliance with the Standard. However, most interactions with App Directories by Desktop Agents will be focused on the retrieval of application details, meaning that these are the priority for standardisation efforts. Further, particular implementors and users of the standard may have different governance or integration requirements for their appD that may be incompatible with providing that endpoint.

Consider downgrading the requirement of this POST endpoint so that is optional by applying the SHOULD or MAY keywords to it in the specification or remove it entirely.

### LINKED ISSUE #567: Add details of where to get more details about an app and where to create an account
## Enhancement Request
While AppD application definitions can provide a variety of details about an application (description, tooltip, images, supported intents  etc.), there are times when further details may be required, particularly if the application is not currently available to / licensed by the user and their firm. In such situations, it would be useful to be able to provide a link to the user to get more info and/or create an account (where that is not possible through the app itself). 

Add support for a 'more info' URL to the appD's application definition format that may be used, by interfaces that provide a UI for browsing applications listed in an AppD, to link the user to additional content about the app.

### Use Case:
- Linking a user to more information about an app, listed in the AppD, that they don't already have access to.

### Additional Information
The Web Applicaiton Manifest format doesn't currently include a similar field, perhaps due to the fact that it is hosted with the application, which (as it is a web application) should be available to serve this purpose on its own. This may not always be the case for financial services apps where they are typically hosted 'on-prem' or installed locally (e.g. native applications).

### LINKED ISSUE #622: Add language metadata to AppD for accessibility
### Use Case:
* An institution has an app that supports multiple languages. They want to include multiple manifest entries, where they have a different version of the app for each language the app supports.
* Consider an implementation of a desktop agent that contains multiple applications across multiple languages. If an end-user is blind and using a screen reader, the screen reader will be unable to independently determine which applications have which languages. The desktop agent will need to provide that language information, which it must determine for a given application.

For example: 
> A bank uses a desktop agent to provide a financial application. The application has an English and Chinese version. The bank wants their English-language employees to be able to use the English version of all of the applications, and their Chinese-language employees to be able to use the Chinese version of the applications. The bank could provide those language identifiers in their manifest, and then the desktop agent could surface only certain languages that end-users require.

> A bank uses a desktop agent to provide financial applications to users in Canada. They provide copies of their applications in French and in English, and may even have language-specific applications. Let's say one of the users at this bank is blind and uses a screen reader. They are a native English speaker, so their operating system, desktop application, and screen reader are all running in English. Whenever the screen reader encounters the title or contents of the _french_ applications, it pronounces the french words as if they are English, resulting in cognitively-taxing confusion. Further more, let's say this user is also fluent in french. With no metadata provided for the french versus english applications, when the user encounters any of the french apps, the screen reader will pronounce french words using english pronunciation rules. In other words, they become gibberish.

### Workflow Description
If the AppD provided metadata for the language of the app (assuming the title of the app and the contents of the app are all in the same language), the desktop agent could surface that language appropriately.

### Additional Information
HTML contains the global attribute, `lang`. This attribute can be used on any HTML element, and would likely be leveraged by any HTML-based desktop agent. The lang attribute typically uses BCP 47, which includes a 2-digit language code (like "en" or "fr") and optionally a 2-digit country code (like "en-US" or "en-GB").
[HTML lang attribute](https://developer.mozilla.org/en-US/docs/Web/HTML/Global_attributes/lang)

The [W3C's Web Application manifest also supports a `lang` attribute](https://w3.org/TR/appmanifest/#lang-member), with [alternative manifest URLs or geolocation and targeting being used to deliver the localized version of the manifest](https://www.w3.org/TR/appmanifest/#internationalization).

### PR DESCRIPTION
**Pull request by kriswest**
Consolidated PR for AppD updates (each of which is raised as a separate PR into this branch for review by AppD discussion group members, maintainers and editors).

resolves #369
resolves #370
resolves #553
resolves #555 
resolves #556 
resolves #558
resolves #560
resolves #561
resolves #562
resolves #565
resolves #566
resolves #567
resolves #622

### CODE REVIEW

**Review by robmoffat — APPROVED** [review]

**Comment by robmoffat** [comment]
Good job - I've approved but I have this one note:

I notice this page hasn't received any changes:  https://deploy-preview-668--lambent-kulfi-cf51a7.netlify.app/docs/next/app-directory/discovery

I was reading this through again, and to my (untrained, noob) eye the part about "Application ID namespace syntax host resolution" really stands out as not relevant now.   If we're moving to the model we talked about on Monday of "get all the appD records from all the app directories" then I don't see what use this naming scheme is.  

The only use-case I can see is one app wanting to load another.. but if that was happening, I don't think the app would know anything about the app directory, would it?

File: CHANGELOG.md [hunk 1]
```diff
@@ -22,6 +22,14 @@ The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/).
 * Added details of FDC3's existing versioning and deprecation policies to the FDC3 compliance page ([#539](https://github.com/finos/FDC3/pull/539))
 * Added a new experimental features policy, which exempts features designated as experimental from the versioning and deprecation policies, to the FDC3 compliance page ([#549](https://github.com/finos/FDC3/pull/549))
 * Add `IntentDeliveryFailed` to the `ResolveError` enumeration to be used when delivery of an intent and context to a targetted app or instance fails. ([#601](https://github.com/finos/FDC3/pull/601))
+* Added `IntentDeliveryFailed` to the `ResolveError` enumeration to be used when delivery of an intent and context to a targetted app or instance fails. ([#601](https://github.com/finos/FDC3/pull/601))
```
Comment (greyseer256) [inline CHANGELOG.md:25]:
```suggestion
* Added `IntentDeliveryFailed` to the `ResolveError` enumeration to be used when delivery of an intent and context to a targeted app or instance fails. ([#601](https://github.com/finos/FDC3/pull/601))
```

Comment (greyseer256) [reply CHANGELOG.md:25]: Also a duplicate of the entry above?

File: src/app-directory/specification/appd.yaml [hunk 2]
```diff
@@ -391,31 +655,118 @@ components:
...
+          description: The name of the intent to 'launch'. In this case the name of an Intent supported by an Application.
+        displayName:
+          type: string
+          description: An optional display name for the intent that may be used in UI instead of the name.
+        contexts:
+          type: array
+          items:
+            type: string
+          description: >-
+            A comma sepaarted list of the types of contexts the intent offered by the application can process.
```
Comment (greyseer256) [inline src/app-directory/specification/appd.yaml:676]:
```suggestion
            A comma separated list of the types of contexts the intent offered by the application can process. 
```

File: src/app-directory/specification/appd.yaml [hunk 3]
```diff
@@ -391,31 +655,118 @@ components:
...
+      properties:
+        alias:
+          type: string
+          description: The Citrix alias / name of the virtual app (passed to the Citrix SelfService qlaunch parameter).
+        arguments:
+          type: string
+          description: Arguments that must be passed on the command line to launch the app in the expected configuration.
+      additionalProperties: false
+    OnlineNativeAppDetails:
+      description: 'Properties used to launch a native apps with `type: onelineNative` that have an online launcher, e.g. online ClickOnce app deployments.'
```
Comment (greyseer256) [inline src/app-directory/specification/appd.yaml:749]:
```suggestion
      description: 'Properties used to launch a native apps with `type: onlineNative` that have an online launcher, e.g. online ClickOnce app deployments.'
```

File: website/static/schemas/next/app-directory.yaml [hunk 4]
```diff
@@ -391,31 +655,118 @@ components:
...
+          description: The name of the intent to 'launch'. In this case the name of an Intent supported by an Application.
+        displayName:
+          type: string
+          description: An optional display name for the intent that may be used in UI instead of the name.
+        contexts:
+          type: array
+          items:
+            type: string
+          description: >-
+            A comma sepaarted list of the types of contexts the intent offered by the application can process.
```
Comment (greyseer256) [inline website/static/schemas/next/app-directory.yaml:676]:
```suggestion
            A comma separated list of the types of contexts the intent offered by the application can process. 
```

File: website/static/schemas/next/app-directory.yaml [hunk 5]
```diff
@@ -391,31 +655,118 @@ components:
...
+      properties:
+        alias:
+          type: string
+          description: The Citrix alias / name of the virtual app (passed to the Citrix SelfService qlaunch parameter).
+        arguments:
+          type: string
+          description: Arguments that must be passed on the command line to launch the app in the expected configuration.
+      additionalProperties: false
+    OnlineNativeAppDetails:
+      description: 'Properties used to launch a native apps with `type: onelineNative` that have an online launcher, e.g. online ClickOnce app deployments.'
```
Comment (greyseer256) [inline website/static/schemas/next/app-directory.yaml:749]:
```suggestion
      description: 'Properties used to launch a native apps with `type: onlineNative` that have an online launcher, e.g. online ClickOnce app deployments.'
```

File: CHANGELOG.md (hunk 1 above)
Comment (kriswest) [reply CHANGELOG.md:25]: yep, fixed

Comment (kriswest) [reply CHANGELOG.md:25]:
```suggestion
```

File: CHANGELOG.md [hunk 6]
```diff
@@ -22,6 +22,14 @@ The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/).
 * Added details of FDC3's existing versioning and deprecation policies to the FDC3 compliance page ([#539](https://github.com/finos/FDC3/pull/539))
 * Added a new experimental features policy, which exempts features designated as experimental from the versioning and deprecation policies, to the FDC3 compliance page ([#549](https://github.com/finos/FDC3/pull/549))
 * Add `IntentDeliveryFailed` to the `ResolveError` enumeration to be used when delivery of an intent and context to a targetted app or instance fails. ([#601](https://github.com/finos/FDC3/pull/601))
```
Comment (kriswest) [inline CHANGELOG.md:24]:
```suggestion
* Add `IntentDeliveryFailed` to the `ResolveError` enumeration to be used when delivery of an intent and context to a targeted app or instance fails. ([#601](https://github.com/finos/FDC3/pull/601))
```

**Comment by kriswest** [comment]
@greyseer256 changes applied

**Comment by kriswest** [comment]
@robmoffat apologies, thought I'd replied to the below:
>I notice this page hasn't received any changes: https://deploy-preview-668--lambent-kulfi-cf51a7.netlify.app/docs/next/app-directory/discovery
>
> I was reading this through again, and to my (untrained, noob) eye the part about "Application ID namespace syntax host resolution" really stands out as not relevant now. If we're moving to the model we talked about on Monday of "get all the appD records from all the app directories" then I don't see what use this naming scheme is.
>
> The only use-case I can see is one app wanting to load another.. but if that was happening, I don't think the app would know anything about the app directory, would it?

The discovery page got merged into the the overview in the formal spec re-org and shuffled around a little so that manual config of appD was promoted to first in the list (as the primary way appDs are discovered IRL). The group hasn't discussed discovery as yet but can look at it in future. Worth noting that the main  use-case for fully qualified appIds is making them globally unique, rather than discovery. However, I think it will remain useful to be able to track back to the appD - perhaps more so once Desktop Agent Bridging arrives and you start seeing comms between apps on different desktop agents, one of whom may not know about the same appDs as the other...

**Review by greyseer256 — APPROVED** [review]

**Pull request merged**